### PR TITLE
feat(PeerReputation): Feature to give reputation to the peer based on its active  connection time

### PR DESCRIPTION
**Pull request by mridullpandey**
https://github.com/paradigmxyz/reth/issues/1052

### CODE REVIEW

**Comment by gakonst** [comment]
Hi! Thank you for the contribution. Please add a PR description and link to the issue this PR closes, so that there's context for us and for everyone following along^^

**Comment by mridullpandey** [comment]
https://github.com/paradigmxyz/reth/issues/1052

File: crates/net/network/src/peers/manager.rs [hunk 1]
```diff
@@ -734,6 +736,8 @@ pub struct Peer {
     kind: PeerKind,
     /// Counts number of times the peer was backed off   
     backoff_counter: u32,
+    ///total time for the connection in active state in milliseconds
+    connection_time: Duration,
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:740]: we need to track an `Instant` here, which is the timestamp of the last tick

File: crates/net/network/src/peers/manager.rs [hunk 2]
```diff
@@ -594,6 +595,7 @@ impl PeersManager {
 
                 peer.state = PeerConnectionState::Out;
                 PeerAction::Connect { peer_id, remote_addr: peer.addr }
+                Peer.connection_time = Peer.connection_time + current.elapsed()
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:598]:
```suggestion
```

File: crates/net/network/src/peers/manager.rs [hunk 3]
```diff
@@ -579,6 +579,7 @@ impl PeersManager {
     fn fill_outbound_slots(&mut self) {
         // as long as there a slots available try to fill them with the best peers
         while self.connection_info.has_out_capacity() {
+            let current = Instant::now()
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:582]: here we need to call the tick function that takes/gets the `current` instant, loops over connected peers, updates the peer's reputation based on `current - last_tick`, updates `last_tick` to `current`

the reputation function should be something that's `f(seconds) = `f((current - last_tick).as_secs())``

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for taking this on,

left some pointers, hope this makes sense.

File: crates/net/network/src/peers/manager.rs (hunk 3 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:582]: Okay sir

File: crates/net/network/src/peers/manager.rs (hunk 2 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:598]: okay sir

File: crates/net/network/src/peers/manager.rs (hunk 1 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:740]: okay I'll do the needful

**Comment by mattsse** [comment]
just following up,

this is already very promising, and mainly the tick function is missing.

we need something like

```rust
fn tick(&mut self, now: Instant) {
  // loop over all _connected_ peers
  // update reputation via seconds connected
     let up_time (now - peer.last_tick).as_secs() as i32;

     peer.reputation + up_time
}
```

**Comment by gakonst** [comment]
@mridullpandey what's your plan for this PR?

**Comment by mridullpandey** [comment]
hey @gakonst  sorry, i'm about to push the changes it's in my local.
]

**Comment by gakonst** [comment]
@mattsse I guess I'm unsure about this PR unless we have a clear way to integrate it in our peer selection algo, i.e. if we give reputation right now it's not really gonna do anything as we only kick peers not reward them

**Comment by mattsse** [comment]
these are separate metrics, this PR is for repupation which is only used to establish or kick peers if they misbehave. And we want to reward peers that we're connected to longer so they're not punished over time if they miss a response here or there.

kicking slowest peers for example is completely different.

File: crates/net/network/src/peers/manager.rs [hunk 4]
```diff
@@ -570,6 +570,21 @@ impl PeersManager {
         }
         Some((*best_peer.0, best_peer.1))
     }
+    //tick function to update reputation of all connected peers
+    fn tick(&mut self, now: Instant) {
+        let mut peers = self.peers.iter_mut().filter(|(_, peer)| {
+                         peer.state.is_connected() &&
+                         !peer.is_banned() &&
+                        (!self.connect_trusted_nodes_only || peer.is_trusted())
+                        })
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:579]: we only need to do this, so checking `is_connected` is sufficient here.

File: crates/net/network/src/peers/manager.rs [hunk 5]
```diff
@@ -736,8 +749,8 @@ pub struct Peer {
     kind: PeerKind,
     /// Counts number of times the peer was backed off   
     backoff_counter: u32,
-    ///total time for the connection in active state in milliseconds
-    connection_time: Duration,
+    ///last tick timestamp of the connection
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:752]:
```suggestion
    /// timestamp of the last tick or when the connection was established.
```

File: crates/net/network/src/peers/manager.rs [hunk 6]
```diff
@@ -570,6 +570,21 @@ impl PeersManager {
...
+        let mut peers = self.peers.iter_mut().filter(|(_, peer)| {
+                         peer.state.is_connected() &&
+                         !peer.is_banned() &&
+                        (!self.connect_trusted_nodes_only || peer.is_trusted())
+                        })
+        // loop over all _connected_ peers
+        for peer in peers {
+        // update reputation via seconds connected
+           let up_time (now - peer.last_tick).as_secs() as i32;
+           peer.reputation + up_time;
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:584]:
```suggestion
           let time_since_last_tick = (now - peer.last_tick).as_secs() as i32;
           peer.reputation += time_since_last_tick;
```

**Review by mattsse — CHANGES_REQUESTED** [review]
getting there, 
few suggestions

File: crates/net/network/src/peers/manager.rs [hunk 7]
```diff
@@ -570,6 +570,21 @@ impl PeersManager {
         }
         Some((*best_peer.0, best_peer.1))
     }
+    //tick function to update reputation of all connected peers
+    fn tick(&mut self, now: Instant) {
```
Comment (gakonst) [inline crates/net/network/src/peers/manager.rs:574]: fmt

File: crates/net/network/src/peers/manager.rs (hunk 4 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:579]: ok

File: crates/net/network/src/peers/manager.rs (hunk 7 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:574]: @gakonst  can you please explain to me about this.

Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:574]: do you mean to format this!!

File: crates/net/network/src/peers/manager.rs (hunk 5 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:752]: okay

File: crates/net/network/src/peers/manager.rs (hunk 6 above)
Comment (mridullpandey) [reply crates/net/network/src/peers/manager.rs:584]: okay

**Comment by gakonst** [comment]
hi @mridullpandey - decided to close this and we'll investigate how to integrate as part of a larger effort on improvements for P2P

**Pull request closed without merging**